### PR TITLE
Session Management with Http Only Cookies

### PR DESCRIPTION
**Pull request by jdorancy-bu**
# Summary

- Refactored white listed path (register, login, logout)
- Use HttpOnly Cookies to share token with the browse and automatically attached for each request. JS cannot touch the token to prevent XSS attack.
- Refactored JWTFilter and added tests
- Added logout

# Development Deployment

Deploying
### Steps to Deploy
- Navigate to the Dev App on Heroku: https://dashboard.heroku.com/apps/trackr-dev/deploy/github
- Go down to **Manual deploy**
- Select your branch from the **Choose a branch to deploy** dropdown
- Click **Deploy Branch**
- Et voila! Your code is route to run in development (assuming your build succeeds).

# Testing

## Manual

- Created user account
- Login/Logout
- Validated requests in network tab of Chrome Dev Tools

## Automated

- Unit tests for JWT Filter
- - Auth with cookie (access and deny)
- - Auth with header (access and deny)
- - Should not filter
- - Should filter

# Code Reviews Checklist
- [X] Manageability
- [X] Architecture
- [X] Maintainability
- [X] Correctness
- [X] Invalid input/states
- [X] Usability
- [X] Reusability
- [X] Object-Oriented Analysis and Design (OOAD) Principles

Reference for the list can be found [here](https://www.codegrip.tech/productivity/the-ultimate-code-review-checklist/?utm_source=website&utm_medium=blog&utm_campaign=best-practices-for-code-review-process)

### CODE REVIEW

**Review by hxb2016 — APPROVED** [review]

File: src/main/java/edu/bu/metcs673/trackr/security/WebSecurityConfig.java [hunk 1]
```diff
@@ -1,72 +1,84 @@
...
+import static edu.bu.metcs673.trackr.common.CommonConstants.JWT_COOKIE_NAME;
+import static javax.servlet.http.HttpServletResponse.SC_OK;
+import static javax.servlet.http.HttpServletResponse.SC_UNAUTHORIZED;
+import static org.springframework.security.config.http.SessionCreationPolicy.STATELESS;
 
 /**
  * Created to specify the spring-security behavior for authorization and
  * authentication related to the JWT tokens
- * 
+ * <p>
```
Comment (tflucker) [inline src/main/java/edu/bu/metcs673/trackr/security/WebSecurityConfig.java:21]: delete the p tag?

File: src/main/java/edu/bu/metcs673/trackr/user/TrackrUserController.java [hunk 2]
```diff
@@ -13,93 +15,106 @@
...
- * @author Tim Flucker
+ * Controller for Users Management. Contains a 'Create' API for new users to register with the
+ * application. Also contains an API which generates a new JWT token for the user to use for
+ * authentication.
  *
+ * @author Tim Flucker
  */
 @Validated
 @RestController
-@RequestMapping("/api/v1/users")
```
Comment (tflucker) [inline src/main/java/edu/bu/metcs673/trackr/user/TrackrUserController.java:34]: why are we getting rid of this? I thought all of our APIs would follow this pattern?

**Review by tflucker — COMMENTED** [review]
Really minor stuff. We can sync up later

File: src/main/java/edu/bu/metcs673/trackr/security/WebSecurityConfig.java (hunk 1 above)
Comment (jdorancy-bu) [reply src/main/java/edu/bu/metcs673/trackr/security/WebSecurityConfig.java:21]: I'll cleanup.

File: src/main/java/edu/bu/metcs673/trackr/user/TrackrUserController.java (hunk 2 above)
Comment (jdorancy-bu) [reply src/main/java/edu/bu/metcs673/trackr/user/TrackrUserController.java:34]: I wanted to create specific paths that are going to be unsecured. For `/api/v1/users` we want to keep it locked down. Trying to do principle of least privilege. If  `/api/v1/users` is unsecured then you  can do the following. 
- `GET /api/v1/users/1`
- `POST GET /api/v1/users/1`
- `DELTE GET /api/v1/users/1`
etc.

**Pull request merged**